### DISCUSSION TITLE
Does task.unique work for pyscript tasks called from hass automation?

### DISCUSSION THREAD

**Discussion by mwright-git** [custom-components/pyscript#217] [Q&A]
I have a pyscript function that gets called from a HASS automation that triggers from a webhook.

From the log.info messages I am seeing, it appears that multiple copies of the pyscript function are running despite there being a task.unique() call at the start of the function.

Am I wrong in my understanding of what task.unique is supposed to do?

Thanks

**Comment by craigbarratt** [comment]
Yes, that is what `task.unique` is meant to do.  The names are unique to the global context, but it sounds like each call to your function should be in the same global context.  Are you just passing a constant, fixed string to `task.unique`?

Also, existing running tasks are not terminated when pyscript scripts are reloaded.  Could you have changed the argument to `task.unique` as you developed your script, but old tasks either didn't call `task.unique` or called it with a different name?

**Reply by mwright-git** [reply]
The parameter to `task.unique` is a fixed string. Below are the first few lines of the task.

    @service
    def Location(parms="null"):
      task.unique("Location")
      log.info(f"Location: parms: {parms}")

I only added the call to `task.unique` very recently and there are no long running copies of the task.

I've found a logfile session from when it went wrong. I've annotated it to indicate which task is which.

    (1) 2021-06-25 14:00:29 INFO (MainThread) [custom_components.pyscript.file.Location.Location] Location: parms: {'command': 'leave', 'area': 'home'}
    (1) 2021-06-25 14:00:29 INFO (MainThread) [custom_components.pyscript.file.Location.Location] ParmTest: Command: leave home
    (1) 2021-06-25 14:00:29 INFO (MainThread) [custom_components.pyscript.file.Location.Location] Command: alarm_control_panel - alarm_arm_away
    (2) 2021-06-25 14:00:32 INFO (MainThread) [custom_components.pyscript.file.Location.Location] Location: parms: {'command': 'enter', 'area': 'home'}
    (2) 2021-06-25 14:00:32 INFO (MainThread) [custom_components.pyscript.file.Location.Location] ParmTest: Command: enter home
    (2) 2021-06-25 14:00:32 INFO (MainThread) [custom_components.pyscript.file.Location.Location] Command: alarm_control_panel - alarm_disarm
    (1) 2021-06-25 14:00:34 INFO (MainThread) [custom_components.pyscript.file.Location.Location] Command (repeat!): alarm_control_panel - alarm_arm_away
    (?) 2021-06-25 14:00:37 INFO (MainThread) [custom_components.pyscript.file.Location.Location] Status: disarmed
    (?) 2021-06-25 14:00:39 INFO (MainThread) [custom_components.pyscript.file.Location.Location] Status: disarmed

**Comment by craigbarratt** [comment]
It would be helpful to see the whole function.  For example, does it make other service calls or run other tasks (which won't be terminated by `task.unique`)?

**Comment by mwright-git** [comment]
OK - I'm not saying the code is particularly good - I'm a bit rusty on coding! I don't think there is anything private here.

```python
@service
def Location(parms="null"):
  task.unique("Location")
  log.info(f"Location: parms: {parms}")
  command = parms["command"]
  location = parms["area"]
  log.info(f"ParmTest: Command: {command} {location}")
  if location != "home": return
  
  entity = "alarm_control_panel"
  
  if command == "enter":
    input_boolean.turn_on(entity_id="input_boolean.mikeathome")
    arm_command = "alarm_disarm"
    expected_state = "disarmed"
  elif command == "leave":
    input_boolean.turn_off(entity_id="input_boolean.mikeathome")
    arm_command = "alarm_arm_away"
    expected_state = "armed_away"
    
  log.info(f"Command: {entity} - {arm_command}")
  
  service.call(entity, arm_command, entity_id="alarm_control_panel.blink_home")
  task.sleep(5)
  if alarm_control_panel.blink_home != expected_state:
    log.info(f"Command (repeat!): {entity} - {arm_command}")
    service.call(entity, arm_command, entity_id="alarm_control_panel.blink_home")
    task.sleep(5)

  log.info(f"Status: {alarm_control_panel.blink_home}")
 ```

**Comment by craigbarratt** [comment]
The code looks good.  I can't see any obvious issue - `task.unique` should certainly stop the prior task.

What versions of HASS / pyscript / python are you using?

It would be good to try a very simple example to see if it fails, eg: do a `task.unique`, `task.sleep` then `log.info`.  Then call it twice in a row.

Also, try turning on pyscript debuggingm eg:
```yaml
logger:
    logs:
        custom_components.pyscript: debug
```

**Reply by mwright-git** [reply]
I'll play with it a bit tomorrow and see if I can isolate the issue.

Versions? 
core-2021.6.6
python 3.8.9
pyscript latest? 1.3.2?
